# 01. Exercise Procedures

## Importing libraries

In [4]:
import pandas as pd
import numpy as np
import os

## importing csv

In [5]:
df_prods=pd.read_csv(r'/Users/nayeraibrahim/Documents/CF assignments/Python/27-10-2020 Instacart Basket Analysis/02 Data/Original Data/products.csv',index_col=False)
df_ords=pd.read_csv(r'/Users/nayeraibrahim/Documents/CF assignments/Python/27-10-2020 Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv',index_col=False)

In [7]:
#create dataframe#
df_test = pd.DataFrame()
#create mixed type column#
df_test['mix'] = ['a', 'b', 1, True]
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
#check mixed type columns#
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
df_test

mix
0     a
1     b
2     1
3  True

In [13]:
#finding nulls#
df_prods.isnull().sum()


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

product_id  product_name  aisle_id  department_id  prices
0           False         False     False          False   False
1           False         False     False          False   False
2           False         False     False          False   False
3           False         False     False          False   False
4           False         False     False          False   False
...           ...           ...       ...            ...     ...
49688       False         False     False          False   False
49689       False         False     False          False   False
49690       False         False     False          False   False
49691       False         False     False          False   False
49692       False         False     False          False   False

[49693 rows x 5 columns]

In [15]:
#creating new dataframe to show null values#
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
#full duplicate rows#
df_dups = df_prods[df_prods.duplicated()]


In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [6]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]


In [7]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [8]:
df_prods_clean.to_csv (r'/Users/nayeraibrahim/Documents/CF assignments/Python/27-10-2020 Instacart Basket Analysis/02 Data/Prepared Data/CHECKED_PRODS.csv')


## Task

In [21]:
#2#
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

the maximum price is unrealistic and this distorted the mean and standard deviation

In [23]:
#3#
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

no mixed-type columns were found

In [24]:
#5#
df_ords.isnull().sum()


Unnamed: 0                      0
order_id                        0
user_identifying_number         0
eval_set                        0
order_number                    0
orders_day_of_week              0
order_hour_of_day               0
days_since_prior_order     206209
dtype: int64

There are 206,209 missing values from the 'days_since_prior_order' column. This indicates that there are no prior orders to the orders placed.

#6# Upon discussion with the client, I'd replace the null values with zero. Since there has been zero days since the prior order.

In [5]:
df_ords_clean = df_ords['days_since_prior_order'].fillna(0)
df_ords_clean

0           0.0
1          15.0
2          21.0
3          29.0
4          28.0
           ... 
3421078    29.0
3421079    30.0
3421080    18.0
3421081     7.0
3421082    30.0
Name: days_since_prior_order, Length: 3421083, dtype: float64

In [26]:
#7#
df_dups = df_ords[df_ords.duplicated()]
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_identifying_number, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

no duplicate values were found

In [7]:
#9#
df_ords_clean.to_csv (r'/Users/nayeraibrahim/Documents/CF assignments/Python/27-10-2020 Instacart Basket Analysis/02 Data/Prepared Data/CHECKED_ORDERS.csv')
